pkg/agent/metrics/prometheus.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/interfacestore"
+	"github.com/vmware-tanzu/antrea/pkg/agent/openflow"
+	"github.com/vmware-tanzu/antrea/pkg/util/env"
+)
+
+type ovsStatManager struct {
+	ofClient     openflow.Client
+	ovsBridge    string
+	ovsTableDesc *prometheus.Desc
+}
+
+func (c *ovsStatManager) getOVSStatistics() (ovsFlowsByTable map[string]float64) {
+	ovsFlowsByTable = make(map[string]float64)
+	flowTableStatus := c.ofClient.GetFlowTableStatus()
+	for _, tableStatus := range flowTableStatus {
+		ovsFlowsByTable[strconv.Itoa(int(tableStatus.ID))] = float64(tableStatus.FlowCount)
+	}
+	return
+}
+
+func (c *ovsStatManager) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.ovsTableDesc
+}
+
+func (c *ovsStatManager) Collect(ch chan<- prometheus.Metric) {
+	ovsFlowsByTable := c.getOVSStatistics()
+	for tableId, tableFlowCount := range ovsFlowsByTable {
+		ch <- prometheus.MustNewConstMetric(
+			c.ovsTableDesc,
+			prometheus.GaugeValue,
+			tableFlowCount,
+			tableId,
+		)
+	}
+}
+
+func newOVSStatManager(ovsBridge string, ofClient openflow.Client) *ovsStatManager {
+	return &ovsStatManager{
+		ofClient:  ofClient,
+		ovsBridge: ovsBridge,
+		ovsTableDesc: prometheus.NewDesc(
+			"antrea_agent_ovs_flow_table",
+			"OVS flow table flow count.",
+			[]string{"table_id"},
+			prometheus.Labels{"bridge": ovsBridge},
+		),
+	}
+}
+
+func InitializePrometheusMetrics(
+	ovsBridge string,
+	ifaceStore interfacestore.InterfaceStore,
+	ofClient openflow.Client) {
+
+	klog.Info("Initializing prometheus metrics")
+	if err := prometheus.Register(prometheus.NewGaugeFunc(
```

Comment:
I remember there was a conversation about using the k8s wrapper objects instead of the Prometheus ones. Has this been abandoned or postponed to a future PR?

Reply:
If postponed, please open an issue - would also be good to explain in the issue why we think the wrappers are superior for our use case

Reply:
The k8s wrapper was proposed by @srikartati and is described in the [this doc](https://github.com/logicalhan/enhancements/blob/1ecb5330ef1254543ac7b462b66f04645f08fc80/keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md).
It offers stability tagging for metrics, e.g mark a metric as stable, alpha, deprecated etc, and offers separate endpoint (e.g /metrics/v1) for the stable metrics.
It would be nice to have as:
A) At some point we could need this functionality.
B) It's a part of k8s ecosystem

However at this point (A) isn't very relevant and Prometheus API definitions as within this patch are still widely common.
I would rather have Prometheus support and handle this in a latter time, when this is a priority.

Reply:
Sounds good to me. Would still like to see a low-priority Github issue with a pointer to the doc.

Reply:
OK

Reply:
See Issue #662